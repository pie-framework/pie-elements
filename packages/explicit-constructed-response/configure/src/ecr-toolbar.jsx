import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import OutlinedInput from '@material-ui/core/OutlinedInput';

const findSlateNode = key => {
  return window.document.querySelector('[data-key="' + key + '"]');
};

class ECRToolbar extends React.Component {
  static propTypes = {
    correctChoice: PropTypes.object,
    classes: PropTypes.object,
    node: PropTypes.object,
    onDone: PropTypes.func,
    onChangeResponse: PropTypes.func.isRequired,
    onToolbarDone: PropTypes.func.isRequired,
    value: PropTypes.shape({
      change: PropTypes.func.isRequired,
      document: PropTypes.shape({
        getNextText: PropTypes.func.isRequired
      })
    })
  };

  state = {};

  componentDidMount() {
    const { correctChoice, node } = this.props;
    const choice = correctChoice || {};

    const domNode = findSlateNode(node.key);

    if (domNode) {
      //eslint-disable-next-line
      const domNodeRect = domNode.getBoundingClientRect();
      const editor = domNode.closest('[data-slate-editor]');
      const editorRect = editor.getBoundingClientRect();
      const top = domNodeRect.top - editorRect.top;
      const left = domNodeRect.left - editorRect.left;

      this.setState({
        markup: choice.label,
        toolbarStyle: {
          position: 'absolute',
          top: `${top + domNodeRect.height + 20}px`,
          left: `${left + 20}px`,
          width: domNodeRect.width
        }
      });
    }
  }

  onDone = () => {
    const { markup: newValue } = this.state;
    const { node, value, onToolbarDone, onChangeResponse } = this.props;
    const update = { ...node.data.toJSON(), value: newValue };
    const change = value.change().setNodeByKey(node.key, { data: update });

    const nextText = value.document.getNextText(node.key);

    change.moveFocusTo(nextText.key, 0).moveAnchorTo(nextText.key, 0);

    onToolbarDone(change, false);
    onChangeResponse(newValue);
  };

  onChange = e => this.setState({ markup: e.target.value });

  render() {
    const { classes } = this.props;
    const { markup, toolbarStyle } = this.state;

    return (
      <div
        style={{
          ...toolbarStyle,
          backgroundColor: '#E0E1E6'
        }}
      >
        <OutlinedInput
          autoFocus
          classes={{
            input: classes.input
          }}
          onChange={this.onChange}
          onBlur={this.onDone}
          value={markup}
        />
      </div>
    );
  }
}

const StyledECRToolbar = withStyles({
  respArea: {
    backgroundColor: '#fff'
  },
  input: {
    backgroundColor: '#fff',
    padding: '10px 20px 10px 10px'
  }
})(ECRToolbar);

export default StyledECRToolbar;
