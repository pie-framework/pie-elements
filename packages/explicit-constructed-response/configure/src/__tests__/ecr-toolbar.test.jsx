import { shallow } from 'enzyme';
import React from 'react';

import { ECRToolbar } from '../ecr-toolbar';

describe('ECRToolbar', () => {
  let onChangeResponse = jest.fn();
  let onToolbarDone = jest.fn();

  const wrapper = () => {
    const defaults = {
      onChangeResponse,
      onToolbarDone,
      classes: {},
      node: {
        key: 1,
        data: {
          get: (prop) => {
            if (prop === 'index') {
              return '2';
            }

            return 'moon';
          },
          toJSON: jest.fn(),
        },
      },
      value: {
        change: jest.fn().mockReturnValue({
          setNodeByKey: jest.fn().mockReturnValue({
            moveFocusTo: jest.fn().mockReturnValue({
              moveAnchorTo: jest.fn(),
            }),
          }),
        }),
        document: {
          getNextText: jest.fn().mockReturnValue({
            key: 1,
          }),
        },
      },
      correctChoice: { value: '0', label: 'moon' },
    };
    const props = { ...defaults };

    return shallow(<ECRToolbar {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    it('onDone: calls onToolbarDone and onChangeResponse', () => {
      const w = wrapper();

      w.instance().onDone();

      expect(onToolbarDone).toBeCalled();
      expect(onChangeResponse).toBeCalled();
    });

    it('onChange', () => {
      const w = wrapper();

      w.instance().onRespAreaChange('test');

      expect(w.instance().state).toEqual({
        markup: '',
        respAreaMarkup: 'test',
        toolbarStyle: {},
      });
    });
  });
});
