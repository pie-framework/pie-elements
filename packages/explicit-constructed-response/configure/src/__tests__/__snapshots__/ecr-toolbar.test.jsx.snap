// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ECRToolbar snapshot renders 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#E0E1E6",
    }
  }
>
  <WithStyles(OutlinedInput)
    autoFocus={true}
    classes={
      Object {
        "input": undefined,
      }
    }
    inputProps={
      Object {
        "maxLength": 25,
      }
    }
    labelWidth={0}
    onBlur={[Function]}
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    value=""
  />
</div>
`;
